File: src/DIY/Prototype/Prototype.java
package DIY.Prototype;

public class Prototype {
    public static void main(String[] args) {
        System.out.println("ORIGINAL CAR:");
        Car car = new Car("Civic", "Base", "2.0 L", "Automatic", "White", "5-door Liftback");
        car.viewSpecifications();

        System.out.println("DUPLICATE CAR: ");
        Dealership dealership = new Dealership();
        dealership.duplicate(car);
        
        System.out.println("CLONED CAR BUT WITH DIFFERENT SPECS: ");
        Component cloned = car.clone();
        ((Car)cloned).setTrim("Type R");
        ((Car)cloned).setBodyType("Coupe");
        ((Car)cloned).setTransmission("Manual");
        ((Car)cloned).setColor("Blue");
        cloned.viewSpecifications();

    }

}

interface Component{
    void viewSpecifications();
    Component clone();
}

class Car implements Component{

    private String model;
    private String trim;
    private String engine;
    private String transmission;
    private String color;
    private String bodyType;
    
    public Car(String model, String trim, String engine,
    String transmission, String color, String bodyType) {
        this.model = model;
        this.trim = trim;
        this.engine = engine;
        this.transmission = transmission;
        this.color = color;
        this.bodyType = bodyType;
    }
    
    @Override
    public Component clone(){
        return new Car(this.model, this.trim, this.engine, this.transmission, this.color, this.bodyType);
    }
    
    @Override
    public void viewSpecifications() {
        System.out.println("Model: " + this.model);
        System.out.println("Trim: " + this.trim);
        System.out.println("Engine: " + this.engine);
        System.out.println("transmission: " + this.transmission);
        System.out.println("bodyType: " + bodyType);
        System.out.println("color: " + this.color + "\n");
    }
    
    public String getTrim() {
        return trim;
    }
    
    public void setTrim(String trim) {
        this.trim = trim;
    }
    
    public String getEngine() {
        return engine;
    }
    
    public void setEngine(String engine) {
        this.engine = engine;
    }
    
    public String getTransmission() {
        return transmission;
    }

    public void setTransmission(String transmission) {
        this.transmission = transmission;
    }
    public String getColor() {
        return color;
    }
    
    public void setColor(String color) {
        this.color = color;
    }
    
    public String getBodyType() {
        return bodyType;
    }
    
    public void setBodyType(String bodyType) {
        this.bodyType = bodyType;
    }
    
    public String getModel() {
        return model;
    }
    public void setModel(String model) {
        this.model = model;
    }
}

class Dealership{
    public void duplicate(Component component){
        Component newComponent = component.clone();
        newComponent.viewSpecifications();
    }
}

File: bin/DIY/Prototype/Prototype.ts
// Prototype.ts

interface Component {
    viewSpecifications(): void;
    clone(): Component;
}

class Car implements Component {
    private model: string;
    private trim: string;
    private engine: string;
    private transmission: string;
    private color: string;
    private bodyType: string;

    constructor(
        model: string,
        trim: string,
        engine: string,
        transmission: string,
        color: string,
        bodyType: string
    ) {
        this.model = model;
        this.trim = trim;
        this.engine = engine;
        this.transmission = transmission;
        this.color = color;
        this.bodyType = bodyType;
    }

    clone(): Component {
        return new Car(
            this.model,
            this.trim,
            this.engine,
            this.transmission,
            this.color,
            this.bodyType
        );
    }

    viewSpecifications(): void {
        console.log(`Model: ${this.model}`);
        console.log(`Trim: ${this.trim}`);
        console.log(`Engine: ${this.engine}`);
        console.log(`transmission: ${this.transmission}`);
        console.log(`bodyType: ${this.bodyType}`);
        console.log(`color: ${this.color}\n`);
    }

    getTrim(): string {
        return this.trim;
    }

    setTrim(trim: string): void {
        this.trim = trim;
    }

    getEngine(): string {
        return this.engine;
    }

    setEngine(engine: string): void {
        this.engine = engine;
    }

    getTransmission(): string {
        return this.transmission;
    }

    setTransmission(transmission: string): void {
        this.transmission = transmission;
    }

    getColor(): string {
        return this.color;
    }

    setColor(color: string): void {
        this.color = color;
    }

    getBodyType(): string {
        return this.bodyType;
    }

    setBodyType(bodyType: string): void {
        this.bodyType = bodyType;
    }

    getModel(): string {
        return this.model;
    }

    setModel(model: string): void {
        this.model = model;
    }
}

class Dealership {
    duplicate(component: Component): void {
        const newComponent = component.clone();
        newComponent.viewSpecifications();
    }
}

// main
(function main() {
    console.log("ORIGINAL CAR:");
    const car = new Car("Civic", "Base", "2.0 L", "Automatic", "White", "5-door Liftback");
    car.viewSpecifications();

    console.log("DUPLICATE CAR: ");
    const dealership = new Dealership();
    dealership.duplicate(car);

    console.log("CLONED CAR BUT WITH DIFFERENT SPECS: ");
    const cloned = car.clone();
    (cloned as Car).setTrim("Type R");
    (cloned as Car).setBodyType("Coupe");
    (cloned as Car).setTransmission("Manual");
    (cloned as Car).setColor("Blue");
    cloned.viewSpecifications();
})();
